auth.go

```diff
@@ -0,0 +1,184 @@
+package pgsrv
+
+import (
+	"bytes"
+	"crypto/md5"
+	"crypto/rand"
+	"fmt"
+)
+
+// authenticator interface defines objects able to perform user authentication
+// that happens at the very beginning of every session.
+type authenticator interface {
+	authenticate() (msg, error)
+}
+
+// authenticationNoPassword responds with auth OK immediately.
+type authenticationNoPassword struct{}
+
+func (*authenticationNoPassword) authenticate() (msg, error) {
+	return authOKMsg(), nil
+}
+
+// messageReadWriter describes objects that handle client-server communication.
+// Objects implementing this interface are used to send password requests to users,
+// and receive their responses.
+type messageReadWriter interface {
+	Write(m msg) error
+	Read() (msg, error)
+}
+
+// passwordProvider describes objects that are able to provide a password given a user name.
+type passwordProvider interface {
+	getPassword(user string) ([]byte, error)
+}
+
+// constantPasswordProvider is a password provider that always returns the same password,
+// which it is given during the initialization.
+type constantPasswordProvider struct {
+	password []byte
+}
+
+func (cpp *constantPasswordProvider) getPassword(user string) ([]byte, error) {
+	return cpp.password, nil
+}
+
+// md5ConstantPasswordProvider is a password provider that returns md5 hash of a given
+// username and a constant password as md5(concat(password, user)).
+type md5ConstantPasswordProvider struct {
+	password []byte
+}
+
+func (cpp *md5ConstantPasswordProvider) getPassword(user string) ([]byte, error) {
+	pu := append(cpp.password, []byte(user)...)
+	puHash := md5.Sum(pu)
+	return puHash[:], nil
+}
+
+// authenticationClearText is an authenticator that requests and accepts a clear text password
+// from the client. It is not recommended to use it for security reasons.
```

Comment:
please rephrase as `authenticationClearText requests and accepts a clear text password...`